Probabilistic assembly scoring based on pairwise interface scores

lafita opened the pull request:
I implemented the assembly scoring as we have discussed in https://github.com/eppic-team/eppic-science/issues/83. The changes can be summarized in:

- A score can be computed for each `Assembly` with the `calcScore()` method. This already accounts for the presence of the different types of symmetries and disjoint subassemblies (I can explain the details if needed). I added a new helper method to the `PowerSet` class.
- `CrystalAssemblies` computes the score for each valid `Assembly` and normalizes them, so that they reflect confidence probabilities. It also assigns the **BIO** call to the highest probability assembly, and **CRYSTAL** to the rest.
- I added another column to the assemblies output file with the score (probability) of each assembly.
- I added the score and confidence to the database upload.

I tried to document the code as much as possible, and I tried to reuse the existing methods and functions where possible, favoring code clarity over performance. If the performance is an issue, I believe we could optimize this part.

The assemblies output for some test cases is the following:

<pre>
# Topologically valid assemblies in 3hbx
 id   Interf cluster ids       Size   Stoichiometry        Symmetry      Score    Predicted by
  1                   {}          1               A              C1       0.00
  2                  {1}          2            A(2)              C2       0.02
  3                  {2}          2            A(2)              C2       0.00
  4                  {4}          3            A(3)              C3       0.00
  5              {1,2,4}          6            A(6)              D3       0.98      eppic,pbd1

# Topologically valid assemblies in 1e94
 id   Interf cluster ids       Size   Stoichiometry        Symmetry      Score    Predicted by
  1                  {1}        1,6          A,A(6)           C1,C6       0.33      eppic,pbd3
  2                   {}        1,1             A,A           C1,C1       0.00
  3                {1,3}        2,6       A(2),A(6)           C2,C6       0.24
  4                {1,5}        2,6       A(2),A(6)           C2,C6       0.09
  5                  {3}        1,2          A,A(2)           C1,C2       0.00
  6                  {5}        1,2          A,A(2)           C1,C2       0.00
  7                {1,4}        6,6       A(6),A(6)           C6,C6       0.06
  8                  {4}        1,6          A,A(6)           C1,C6       0.00
  9            {1,3,4,5}       6,12      A(6),A(12)           C6,D6       0.14
 10              {1,3,4}       6,12      A(6),A(12)           C6,D6       0.05
 11              {1,3,5}       6,12      A(6),A(12)           C6,D6       0.06
 12              {1,4,5}       6,12      A(6),A(12)           C6,D6       0.02
 13              {3,4,5}       1,12         A,A(12)           C1,D6       0.00       pbd1,pbd2
 14                {3,4}       1,12         A,A(12)           C1,D6       0.00
 15                {3,5}       1,12         A,A(12)           C1,D6       0.00
 16                {4,5}       1,12         A,A(12)           C1,D6       0.00

# Topologically valid assemblies in 4cqm
 id   Interf cluster ids       Size   Stoichiometry        Symmetry      Score    Predicted by
  1                  {2}      2,1,1        A(2),A,A        C2,C1,C1       1.00           eppic
  2                   {}      1,1,1           A,A,A        C1,C1,C1       0.00 pbd1,pbd2,pbd3,pbd4

</pre>

- **3HBX** is a D3 tetramer with an induced interface (#3) that is correctly identified. 
- **1E94** is a disjoint assembly (co-crystalization) with borderline (low confidence) interface calls. We can see how this uncertainty is translated into the assembly prediction, with the highest call being 0.33 and two more assemblies with scores higher than 0.1, so it suggests more evidence is needed to make a reliable decision.
- Finally, I have encountered some examples where the interface grouping (clustering) is not correct, like **4CQM**, and the assembly generation is thus also incorrect (I will open a separate issue). This scoring procedure assumes that the clustering is correct, so that could be a caveat in the confidence calibration (we would have to benchmark it).

josemduarte commented:
Regarding your point with incorrect interface clustering for 4cqm, you are totally right: if we get the clustering wrong then all assembly calculations downstream from there will be incorrect, including finding which assemblies are valid or scoring them. If you can add an issue about that particular bad clustering case that would be great, thanks!

lafita commented:
I have dropped the score normalization (normalization can be easily added when doing the benchmarking from the unnormalized scores) and added a warning when the sum of probabilities of the valid assemblies is low (indicating a problem with ennumeration).

I have also added a maximum number of engaged interfaces for the score calculation and if it is exceeded, a warning is shown and interfaces are removed from the score calculation, starting from the smallest probability ones, until the threshold of 10 interfaces is reached. This will minimally affect the score except for cases where there are more than 10 meaningful interface clusters contributing to the score (even possible).

lafita commented on eppic-cli/src/main/java/eppic/assembly/Assembly.java:
```diff
@@ -955,21 +974,64 @@ public String toString() {
+			if (pa.getAssemblyGraph().getSubAssemblies().size() == 
+					this.getAssemblyGraph().getSubAssemblies().size())
+				pss.add(ps);
+		}
```
The condition is needed because only those parents that form the same assembly have to be included in the probability of the assembly. This is checked by removing interfaces and seeing if the components of the graph are the same (removing edges can only increase the number of components, in which case that parent does not have to be included).

As an example, a dihedral assembly with interfaces 1,2 and 3, would have contributions from subsets (1,2), (2,3) and (1,3), but not from (1), (2) or (3), because those form dimers.

It is also a way to generically calculate the minimum number of interfaces required for an assembly to be formed. For homomeric dihedrals is 2 and for cyclics 1, but it gets more complicated for heteromers.

josemduarte replied:
Thanks for the explanation! it's clear now. That's a nice solution to the problem

lafita commented on eppic-cli/src/main/java/eppic/assembly/Assembly.java:
```diff
@@ -955,21 +974,64 @@ public String toString() {
+		for (PowerSet ps : pss) {
+			double prob = 1;
+			for (int i = 1; i < ps.size()+1; i++) {
+				double p = iecl.getCombinedClusterPredictor(i).getScore();
```
True, this was one of my questions in #147. If you have the right way to do it, we can update the code.
